### PR TITLE
Safer Wait for Snapshot Success in ClusterPrivilegeTests

### LINKED ISSUE #38030: ClusterPrivilegeTests.testThatSnapshotAndRestore failing in master
`
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/1650/console
`

Does not reproduce for me locally
```
REPRODUCE WITH: ./gradlew :x-pack:plugin:security:unitTest \
  -Dtests.seed=1E58C34D450B1482 \
  -Dtests.class=org.elasticsearch.integration.ClusterPrivilegeTests \
  -Dtests.method="testThatSnapshotAndRestore" \
  -Dtests.security.manager=true \
  -Dtests.locale=en-CA \
  -Dtests.timezone=America/Adak \
  -Dcompiler.java=11 \
  -Druntime.java=8
```

### PR DESCRIPTION
**Pull request by original-brownbear**
* The existence of a snapshot in the CS with state `SUCCESS` means that there is still a snapshot ongoing (it must still be finalized in the repository). Hence this change translates the `SUCCESS`  state in the CS to in progress on the transport layer.
* closes #38030

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

File: x-pack/plugin/security/src/test/java/org/elasticsearch/integration/ClusterPrivilegeTests.java [hunk 1]
```diff
@@ -203,5 +205,12 @@ private void waitForSnapshotToFinish(String repo, String snapshot) throws Except
             SnapshotsStatusResponse response = client().admin().cluster().prepareSnapshotStatus(repo).setSnapshots(snapshot).get();
             assertThat(response.getSnapshots().get(0).getState(), is(SnapshotsInProgress.State.SUCCESS));
         });
+        // The status of the snapshot in the repository can become SUCCESS before it is fully finalized in the cluster state so wait for
+        // it to disappear from the cluster state as well
+        assertBusy(() -> {
+            SnapshotsInProgress response =
+                client().admin().cluster().state(new ClusterStateRequest()).get().getState().custom(SnapshotsInProgress.TYPE);
+            assertThat(response.entries(), Matchers.empty());
+        });
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/integration/ClusterPrivilegeTests.java]: @original-brownbear you've got my rubber stamp LGTM , but shouldn't we fix the `prepareSnapshotStatus(repo).setSnapshots(snapshot)` API to not return `SUCCESS` if there is still an `SnapshotsInProgress.TYPE` entry?

**Review by albertzaharovits — DISMISSED** [review]
LGTM but I think the snapshot status API needs fix so that it returns success consistently with respect to the `SnapshotsInProgress.TYPE` metadata entry. If, for some obscure reason this is not possible, I believe a comment in the main code (similar to the one here) is required. If/when the change in main lands, I would remove this extra condition here.

**Comment by original-brownbear** [comment]
@albertzaharovits thanks for taking a look :) You know ... I just realized these tests run against a single node => we shouldn't have any races here with the state like that => I'm double checking the solution, sorry for the noise!

**Comment by original-brownbear** [comment]
Ok, this is a mess :) I was able to fix the API (at least for the single node case) via https://github.com/elastic/elasticsearch/pull/40943/commits/f7c4afb0bc8a027ec292264967b46ce051557199

We have two APIs for getting the current list of snapshots, `org.elasticsearch.snapshots.SnapshotsService#currentSnapshots(java.lang.String)` and `org.elasticsearch.snapshots.SnapshotsService#currentSnapshots(java.lang.String, java.util.List<java.lang.String>)`. 
* One of them returns `org.elasticsearch.snapshots.SnapshotInfo` and correctly translates a `SUCCESS` state in the cluster state as still technically `IN_PROGRESS` and sets that state on the `SnapshotInfo`.
* The other one (used by `org.elasticsearch.action.admin.cluster.snapshots.status.TransportSnapshotsStatusAction` which is what this test uses) returns the raw `SnapshotsInProgress.Entry` list, so `SUCCESS` stays `SUCCESS` when in actuality the existence of a `SUCCESS` entry means that we still have to finalize the snapshot in the repository potentially. Due to the fact that the calls to `SnapshotService` are used in multiple places, I decided to for now just hack the fix of translating success for ongoing snapshots to "in progress/started" into the transport API that isn't used by the snapshot process itself. This fixed the test broken here for me and also makes the REST API a little more useful :)

**Comment by albertzaharovits** [comment]
@original-brownbear If I understand it correctly, every `SnapshotsInProgress.Entry` entry is a sign that the snapshot is "in-progress", ie `SnapshotState.IN_PROGRESS` or `SnapshotsInProgress.State.STARTED` or `SnapshotsInProgress.State.SUCCESS` mean the same thing from the "on-progress" perspective.

I would defer to others to lean on the API change. If you don't mind, I'll shirk away from the reviewers group.
If I were you, I would've done the `state == SnapshotsInProgress.State.SUCCESS ? SnapshotsInProgress.State.STARTED : state` hack on the client side, even though it's apparently harder. For example, if we go with this proposed API change I think we need an accompanying test of its own. And the PR label needs to be changed as well, and probably further backporting the change. Again, I will leave to others for authoritative suggestions.

But, can't the test(s) use `TransportGetSnapshotsAction`, for "in-progress" polling, which internally uses the "correctly" functioning `SnapshotService#currentSnapshots` ?

**Comment by original-brownbear** [comment]
@albertzaharovits 

> I would defer to others to lean on the API change. If you don't mind, I'll shirk away from the reviewers group.

Sure no worries, thanks for the input so far!

> But, can't the test(s) use TransportGetSnapshotsAction, for "in-progress" polling, which internally uses the "correctly" functioning SnapshotService#currentSnapshots ?

Yea, but I figured we're seeing an actual bug in the API here, so why not fix it. We don't have other code dependent on this API it seems and it's just a 3 line change to get rid of a mean inconsistency here :)

>  hack on the client side, even though it's apparently harder.

I don't think we can really do anything on the client side here. To the client, the reason for `SUCCESS` is not visible (it doesn't know if the snapshot state was read from the repo or cluster state) -> it can't translate `SUCCESS` conditionally.

**Comment by original-brownbear** [comment]
@andrershov @DaveCTurner can one of you take a look here, this should be a fairly trivial and easy one :)

**Review by andrershov — COMMENTED** [review]
I don't like that hack. 
The issue arises from the fact, that we're reusing SnapshotInProgress.State enum when returning a list of snapshots, which can contain not only snapshots in progress but also snapshots that are not currently running.
Can we look into creating a new enum that describes state not only of running snapshots but for snapshots that are not running as well?

**Comment by original-brownbear** [comment]
@andrershov 

> Can we look into creating a new enum that describes state not only of running snapshots but for snapshots that are not running as well?

We already have that with `org.elasticsearch.snapshots.SnapshotState`, but the problem is that the current response here simply doesn't use that. I'd rather not change the serialization format for this kind of trivial thing though just to fix the API.
Also, we're doing this hack in multiple places (the logic of interpreting state for snapshots from the CS and from the repo differently is duplicated multiple times, unfortunately), we're probably better off first drying this up and then moving these responses to the new Enum.

**Comment by andrershov** [comment]
As a quick-and-dirty fix, I think your solution is ok, but can you open a new GitHub issue to re-think the API of getting snapshot status and mark it with team-discuss?

**Comment by original-brownbear** [comment]
> This PR #5123 has introduced TransportSnapshotsStatusAction. The PR is titled "Add ability to get snapshot status for running snapshots" and I'd expect that only running snapshots are included in the output, but currently we're including information about non-running snapshots as well.
Probably there is already an action that could be used to get the information about all the snapshots (not detailed information). And if you want to, you may use TransportSnapshotsStatusAction to get detailed information about the snapshots that are currently running.

> As a quick-and-dirty fix, I think your solution is ok, but can you open a new GitHub issue to re-think the API of getting snapshot status and mark it with team-discuss?

Yea I agree with all points in the above + we def. should have an issue open about this => I'll add one :)

**Comment by original-brownbear** [comment]
There's probably also gonna be some overlap with the snapshot life cycle management effort and this discussion, so link https://github.com/elastic/elasticsearch/issues/41210 :)

**Comment by original-brownbear** [comment]
@ywelsch could you have a look here please? Thanks!

**Comment by ywelsch** [comment]
Reusing the same enum for snapshots that are in progress and those that have completed has created this odd behavior where users can't distinguish between an in-progress snapshot that's in its finalization step and one where finalization was successfully completed. The same applies to the FAILED state as well. Even worse, we reuse the same enum across shard snapshot status and global snapshot status, which has caused even more confusion. I think it's a good time now to start separating these instead of adding a workaround. We should fix the test first without making changes to ES so that we can get test coverage back. Waiting for successful snapshot can be done by having the test first check whether `GET /_snapshot/my_backup/_current`  returns a `SnapshotMissingException` and then do the snapshot status check.

**Comment by original-brownbear** [comment]
@ywelsch alright, no hack it is :) I just reverted to my initial check here and simply made the test verify that the snapshot is gone from the CS like we do in other tests.

Can you take another look?

**Comment by original-brownbear** [comment]
@ywelsch ping :)

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
thanks @ywelsch !

**Pull request merged**